### ISSUE TITLE
Source files are modified on import

### ISSUE THREAD

**Issue by ethack** [jmathai/elodie#252]
TL;DR Should [this line](https://github.com/jmathai/elodie/blob/fd3cab4035dca583fe512247018f5180a012aacf/elodie/filesystem.py#L336) be `media.get_original_name()` instead?  What follows below is the behavior I'm seeing and the process I went through in case I'm wrong with the suggested fix.

-----

I am seeing the same or a similar issue as #237 on the current master branch.  But I have another possible explanation: exiftool is modifying the original file.

[This line](https://github.com/jmathai/elodie/blob/fd3cab4035dca583fe512247018f5180a012aacf/elodie/media/media.py#L54) passes the `-overwrite_original` flag to exiftool.  This is updating the date modified and changing the file's checksum.  I can prevent the date modified from changing by adding a `-preserve` flag to that same list in the media.py file.  But this still modifies the original file contents.  If I remove the `-overwrite_original` flag from elodie then exiftool creates a backup of the original file as described in its documentation.

With the following config file:
```
[Directory]
year=%Y
month=%m
full_path=%year/%month
```

* Some files are sorted into year/month based on their original date modified time.  This is desired behavior.  I'm guessing these are files with exif date data already in the file.
* Some files are sorted into year/month based on the new date modified time.  This is not desired behavior. I'm guessing these are files with no exif date data.

In the screenshot below:
* unsorted_times_original - Contains the original files untouched by elodie.
* unsorted_times - Was a copy made of unsorted_times_original.  This was the source directory given to elodie to sort.
* sorted - Is the resulting directory structure after unsorted_times was sorted.

<img width="574" alt="image" src="https://user-images.githubusercontent.com/1696711/31159796-36761b42-a891-11e7-8cd0-aaa3c3727a43.png">

You can see that the file times were modified and that about 3KB in size was added to each file.  The hashes contained in hashes.json match the modified files and not the originals.  I assumed that hashes.json was so that if elodie ran across the same source file again it would know it had already processed it and could ignore it.  If that's the case then the hashes would never match new source files because the hash is of a file that elodie modified.

I stepped through the code and determined where the original file was being modified (at least for the first time).  [This line](https://github.com/jmathai/elodie/blob/fd3cab4035dca583fe512247018f5180a012aacf/elodie/media/media.py#L285) is when the file is modified and I've included a call stack output below immediately after that line is executed in case it helps.

```
File "elodie.py", line 345, in <module>
  main()
File "elodie/venv/lib/python2.7/site-packages/click/core.py", line 716, in __call__
  return self.main(*args, **kwargs)
File "elodie/venv/lib/python2.7/site-packages/click/core.py", line 696, in main
  rv = self.invoke(ctx)
File "elodie/venv/lib/python2.7/site-packages/click/core.py", line 1060, in invoke
  return _process_result(sub_ctx.command.invoke(sub_ctx))
File "elodie/venv/lib/python2.7/site-packages/click/core.py", line 889, in invoke
  return ctx.invoke(self.callback, **ctx.params)
File "elodie/venv/lib/python2.7/site-packages/click/core.py", line 534, in invoke
  return callback(*args, **kwargs)
File "elodie.py", line 115, in _import
  trash, allow_duplicates)
File "elodie.py", line 64, in import_file
  media, allowDuplicate=allow_duplicates, move=False)
File "elodie/elodie/filesystem.py", line 336, in process_file
  media.set_original_name()
File "elodie/elodie/media/media.py", line 256, in set_original_name
  status = self.__set_tags(tags)
File "elodie/elodie/media/media.py", line 288, in __set_tags
  return status != ''
```

Issue #237 already established that the modified time of the file should not change.  I would add that I do not wish the original files to be modified either.  I started out my organization process by doing a  checksum of all my source files (plus some other metadata) so that I could reference in case my prior folder structure was useful for organizing certain files later.  If elodie modifies the original file then all of those checksums will be invalid.

**Comment by jmathai** [comment]
Thanks so much for the detailed report. Can you paste in the command you're running in its entirety?

I am wondering if there's a code path we don't have tests for. [This line](https://github.com/jmathai/elodie/blob/master/elodie/tests/filesystem_test.py#L444), for example, is verifying that the checksum of the original file and the destination file are identical. And that test passes.

**Comment by ethack** [comment]
Absolutely.  And in the back of my mind this whole time has been the question that I'm doing something wrong and shooting my own foot :)

```
python elodie.py import --source ~/Pictures/unsorted_times/ --destination ~/Pictures/sorted
```

I'm running elodie from in a python virtualenv.  I'm on a clean up-to-date master branch (git pull and git status return no changed files).  I'm using python version 2.7.13 and exiftool version 10.55.

I just installed the test dependencies and ran `nosetests` from the repo root.  That test actually fails for me, along with some others.  I've attached the entire output in a separate file.  I should note that the tests took a long time and I went to run elodie manually at the same time so I deleted the hash.json file.  After I deleted that file the tests finished.  I don't know if that was coincidence or if by deleting the hash.json file I would have messed up some of the tests, but I thought it was worth mentioning.

```
[...]
======================================================================
FAIL: elodie.tests.filesystem_test.test_process_file_plain
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ethan/Documents/elodie/venv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/ethan/Documents/elodie/elodie/tests/filesystem_test.py", line 445, in test_process_file_plain
    assert helper.path_tz_fix(os.path.join('2015-12-Dec','Unknown Location','2015-12-05_00-59-26-photo.jpg')) in destination, destination
AssertionError: /var/folders/lb/8glnrd7s0_931p9vrn9_5c8m0000gn/T/2015/12/2015-12-05_00-59-26-photo.jpg
-------------------- >> begin captured logging << --------------------
root: DEBUG: [u'exiftool', u'-stay_open', u'True', u'-@', u'-', u'-common_args', u'-G', u'-n', '-overwrite_original', u'-config', u'"/Users/ethan/Documents/elodie/configs/ExifTool_config"']
root: DEBUG: [u'exiftool', u'-stay_open', u'True', u'-@', u'-', u'-common_args', u'-G', u'-n', '-overwrite_original', u'-config', u'"/Users/ethan/Documents/elodie/configs/ExifTool_config"']
root: DEBUG: [u'exiftool', u'-stay_open', u'True', u'-@', u'-', u'-common_args', u'-G', u'-n', '-overwrite_original', u'-config', u'"/Users/ethan/Documents/elodie/configs/ExifTool_config"']
root: DEBUG: [u'exiftool', u'-stay_open', u'True', u'-@', u'-', u'-common_args', u'-G', u'-n', '-overwrite_original', u'-config', u'"/Users/ethan/Documents/elodie/configs/ExifTool_config"']
root: DEBUG: [u'exiftool', u'-stay_open', u'True', u'-@', u'-', u'-common_args', u'-G', u'-n', '-overwrite_original', u'-config', u'"/Users/ethan/Documents/elodie/configs/ExifTool_config"']
root: DEBUG: [u'exiftool', u'-stay_open', u'True', u'-@', u'-', u'-common_args', u'-G', u'-n', '-overwrite_original', u'-config', u'"/Users/ethan/Documents/elodie/configs/ExifTool_config"']
root: DEBUG: [u'exiftool', u'-stay_open', u'True', u'-@', u'-', u'-common_args', u'-G', u'-n', '-overwrite_original', u'-config', u'"/Users/ethan/Documents/elodie/configs/ExifTool_config"']
root: DEBUG: [u'exiftool', u'-stay_open', u'True', u'-@', u'-', u'-common_args', u'-G', u'-n', '-overwrite_original', u'-config', u'"/Users/ethan/Documents/elodie/configs/ExifTool_config"']
--------------------- >> end captured logging << ---------------------
[...]
Ran 228 tests in 1609.700s

FAILED (SKIP=1, failures=22)
```

[elodie_test_output.txt](https://github.com/jmathai/elodie/files/1356361/elodie_test_output.txt)

I verified that changing the line to media.get_original_name() left the original files untouched.

I've attached the sample files I'm working with as well.
[unsorted_times_original.zip](https://github.com/jmathai/elodie/files/1356223/unsorted_times_original.zip)

**Comment by jmathai** [comment]
Did you have a customized `config.ini` when you ran the tests? It appears that might be why the tests failed. Could you replace that file with `config.ini-sample`, add a mapquest key,  and run the tests again.

**Comment by ethack** [comment]
You are right, I was using the one from above.  With a default `config.ini` file, a valid mapquest key, and clean master branch all tests passed (besides the one that is skipped).

FWIW, I'm seeing the same issue with the source files being modified with the default `config.ini` file as well.

Are you able to reproduce the issue with the files I sent?  If not, I'll have to try and work out what's special about my environment.

**Comment by mattca** [comment]
Looks like I am also seeing this issue. Python 2.7.13 / exiftool 10.55.

**Comment by kmajzlik** [comment]
This program changed file dates on all my photos 👎

**Comment by jmathai** [comment]
Lost track of this bug. I tried to reproduce this with @ethack's sample photos but was not able to. @kmajzlik could you provide sample files and the import command you ran?

I'm on commit 8950f75.

```zsh
(elodie) ➜  elodie git:(master) ✗ cat .git/refs/heads/master
8950f756fe03774d8a34e6fd997372328c056438
```

Here is my update command.

```zsh
(elodie) ➜  elodie git:(master) ✗ ./elodie.py import --destination=/Users/jaisen/Downloads/etest /Users/jaisen/Downloads/unsorted_times_original
/Users/jaisen/Downloads/unsorted_times_original/CDGrQ.jpg -> /Users/jaisen/Downloads/etest/2018-09-Sep/Unknown Location/2018-09-17_05-29-39-cdgrq.jpg
/Users/jaisen/Downloads/unsorted_times_original/0912091853a.jpg -> /Users/jaisen/Downloads/etest/2009-09-Sep/Unknown Location/2009-09-13_07-53-27-0912091853a.jpg
****** SUMMARY ******
Metric      Count
--------  -------
Success         2
Error           0
```

I ran an md5 on the source and imported files and they matched.

```zsh
# for 0912091853a.jpg
(elodie) ➜  elodie git:(master) ✗ md5 /Users/jaisen/Downloads/unsorted_times_original/0912091853a.jpg
MD5 (/Users/jaisen/Downloads/unsorted_times_original/0912091853a.jpg) = 1c45a93eecd8b720d34318e71c79e6ae
(elodie) ➜  elodie git:(master) ✗ md5 /Users/jaisen/Downloads/etest/2009-09-Sep/Unknown\ Location/2009-09-13_07-53-27-0912091853a.jpg
MD5 (/Users/jaisen/Downloads/etest/2009-09-Sep/Unknown Location/2009-09-13_07-53-27-0912091853a.jpg) = 1c45a93eecd8b720d34318e71c79e6ae

# for CDGrQ.jpg
(elodie) ➜  elodie git:(master) ✗ md5 /Users/jaisen/Downloads/unsorted_times_original/CDGrQ.jpg
MD5 (/Users/jaisen/Downloads/unsorted_times_original/CDGrQ.jpg) = cc80794d85fded64530d5ab16b817b64
(elodie) ➜  elodie git:(master) ✗ md5 /Users/jaisen/Downloads/etest/2018-09-Sep/Unknown\ Location/2018-09-17_05-29-39-cdgrq.jpg
MD5 (/Users/jaisen/Downloads/etest/2018-09-Sep/Unknown Location/2018-09-17_05-29-39-cdgrq.jpg) = cc80794d85fded64530d5ab16b817b64
```

**Comment by jmathai** [comment]
I think I was able to reproduce this. I unzipped the @ethack's file and ran the md5 and got a mismatch.

I'll need to look into this a bit more now that I've reproduced it. There are [tests](https://github.com/mattca/elodie/commit/f80f6629b99481f0d261323b4f54e66a838f4930#diff-d2de6d320f1b7278ef0d36c1418bfc2cR513 to verify that the source file's md5 does not change.

**Comment by alex-phillips** [comment]
I'm getting several md5 changes as well between verify runs. This may be related. Any update on this?

**Comment by tjaadvd** [comment]
I am running into this isue as well. On a Macbook Pro, clean install and added mapquest api key.

**Comment by alex-phillips** [comment]
@tjaadvd in the meantime, i've been using https://github.com/ambv/bitrot if that helps

**Comment by jmathai** [comment]
Kudos to @mattca for submitting a PR (#272) for this which has now been merged. 👏
